### PR TITLE
various fixes and tweaks for integer parsing

### PR DESCRIPTION
**Pull request by poettering**
A follow-up for #15991. Addressing #15985

### CODE REVIEW

File: src/basic/parse-util.c [hunk 1]
```diff
@@ -359,20 +359,34 @@ int safe_atou_full(const char *s, unsigned base, unsigned *ret_u) {
...
+
+        if (FLAGS_SET(base, SAFE_ATO_REFUSE_LEADING_WHITESPACE) &&
+            strchr(WHITESPACE, s[0]))
+                return -EINVAL;
 
         s += strspn(s, WHITESPACE);
 
+        if (FLAGS_SET(base, SAFE_ATO_REFUSE_PLUS_MINUS) &&
+            IN_SET(s[0], '+', '-'))
+                return -EINVAL; /* Not that we check the "-" prefix again a second time below, but return a
```
Comment (keszybz) [inline src/basic/parse-util.c]: Not that it matters, but "note", "I.e.".

File: src/basic/parse-util.c [hunk 2]
```diff
@@ -359,20 +359,34 @@ int safe_atou_full(const char *s, unsigned base, unsigned *ret_u) {
...
+                                 * different error. i.e. if the SAFE_ATO_REFUSE_PLUS_MINUS we blanket refuse
+                                 * +/- prefixed integers, while if it is missing we'll just return ERANGE,
+                                 * because the string actually parses correctly, but doesn't fit in the
+                                 * return type. */
+
+        if (FLAGS_SET(base, SAFE_ATO_REFUSE_LEADING_ZERO) &&
+            s[0] == '0' && s[1] != 0)
```
Comment (keszybz) [inline src/basic/parse-util.c]: This will let '001' through.

File: src/basic/parse-util.c [hunk 3]
```diff
@@ -480,13 +507,24 @@ int safe_atou16_full(const char *s, unsigned base, uint16_t *ret) {
...
+            IN_SET(s[0], '+', '-'))
+                return -EINVAL;
+
+        if (FLAGS_SET(base, SAFE_ATO_REFUSE_LEADING_ZERO) &&
+            s[0] == '0' && s[1] != 0)
+                return -EINVAL;
+
         errno = 0;
-        l = strtoul(s, &x, base);
+        l = strtoul(s, &x, SAFE_ATO_MASK_FLAGS(base));
```
Comment (keszybz) [inline src/basic/parse-util.c:568]: Hmm, why do we need to check those conditions above, if we pass those flags to `strtoul`? I guess there's some reason for it, so maybe add a comment.

File: src/test/test-user-util.c [hunk 4]
```diff
@@ -54,13 +54,41 @@ static void test_parse_uid(void) {
...
+
+        r = parse_uid(" 1234", &uid);
+        assert_se(r == -EINVAL);
+        assert_se(uid == 100);
+
         r = parse_uid("01234", &uid);
-        assert_se(r == 0);
-        assert_se(uid == 1234);
+        assert_se(r == -EINVAL);
+        assert_se(uid == 100);
```
Comment (keszybz) [inline src/test/test-user-util.c:95]: Please also add a test with multiple leading zeros, e.g. `001`, `00`, `000`.

File: src/basic/parse-util.c (hunk 2 above)
Comment (poettering) [reply src/basic/parse-util.c]: ouch

File: src/basic/parse-util.c (hunk 3 above)
Comment (poettering) [reply src/basic/parse-util.c:568]: we don't pass them to strtoul(). The`SAFE_ATO_MASK_FLAGS()` macro *masks* the flags and leaves only the base in the value

File: src/basic/parse-util.c (hunk 2 above)
Comment (poettering) [reply src/basic/parse-util.c]: hmm, actually it does not. If the `SAFE_ATO_REFUSE_LEADING_ZERO` flag is set, then for `001` the `s[0] == '0'` test will hold. but the second test `s[1] != 0` will hold too, and EINVAL is returned.

Maybe the confusion is from the fact that `0` and `'0'` look so similar?

I'll reword this as:

`s[0] && !streq(s, "0")`

That should make things clearer, no?

**Comment by poettering** [comment]
Force pushed new version. Only the issues pointed out addressed. But see comment above. PTAL

File: src/basic/parse-util.c (hunk 2 above)
Comment (0x2b3bfa0) [reply src/basic/parse-util.c]: `s[0] && !streq(s, "0")` is true for any non-empty string but "0", however your [last commits](https://github.com/systemd/systemd/pull/16033/files#diff-327855c3af836dac88385f2811f3929dR408) does it right.

File: src/basic/parse-util.c (hunk 3 above)
Comment (0x2b3bfa0) [reply src/basic/parse-util.c:568]: Is this approach clean enough?

Comment (keszybz) [reply src/basic/parse-util.c:568]: Oh, OK. I missed the inversion in the macro.

File: src/basic/parse-util.c [hunk 5]
```diff
@@ -70,26 +70,24 @@ int parse_pid(const char *s, pid_t* ret_pid) {
...
-        long l;
+        unsigned m;
+        int r;
 
         assert(s);
-        assert(ret);
 
-        s += strspn(s, WHITESPACE);
-        if (s[0] == '-')
-                return -ERANGE;
```
Comment (keszybz) [inline src/basic/parse-util.c:81]: I don't think we should allow leading whitespace. Our parsers generally strip whitespace earlier, so if we get here with some leading whitespace, something strange is going on. But if the user somehow managed to feed us leading whitespace, we should fail here. But this behaviour didn't change with this PR, so let's leave that for later.

File: src/basic/user-util.c [hunk 6]
```diff
@@ -74,22 +74,36 @@ int parse_uid(const char *s, uid_t *ret) {
...
+                /* Single number with no dash. */
+                u = l;
+        else if (!*s)
+                /* Trailing dash is an error. */
+                return -EINVAL;
+        else {
+                r = parse_uid(s, &u);
+                if (r < 0)
+                        return r;
+        }
```
Comment (keszybz) [inline src/basic/user-util.c:109]: We don't have any tests for parse_uid_range() :(

File: src/basic/parse-util.c [hunk 7]
```diff
@@ -402,13 +430,17 @@ int safe_atou_full(const char *s, unsigned base, unsigned *ret_u) {
 }
 
 int safe_atoi(const char *s, int *ret_i) {
+        unsigned base = 0;
         char *x = NULL;
         long l;
 
         assert(s);
 
+        s += strspn(s, WHITESPACE);
```
Comment (keszybz) [inline src/basic/parse-util.c:439]: Is this a good idea? Why should we allow leading whitespace here?

File: src/basic/parse-util.c [hunk 8]
```diff
@@ -459,13 +493,17 @@ int safe_atollu_full(const char *s, unsigned base, long long unsigned *ret_llu)
 }
 
 int safe_atolli(const char *s, long long int *ret_lli) {
+        unsigned base = 0;
         char *x = NULL;
         long long l;
 
         assert(s);
 
+        s += strspn(s, WHITESPACE);
```
Comment (keszybz) [inline src/basic/parse-util.c:502]: Ditto.

File: src/basic/parse-util.c (hunk 7 above)
Comment (poettering) [reply src/basic/parse-util.c:439]: well, not sure if its, but strtol() does it implicitly when parsing normally, and that's documented. I wanted to make this as little invasive as possible, so I tried to add the 0b + 0o support in a fashion that behaves as similar as possible to how 0x is already handled.

(I must say, given all the weird behaviour of strtol() that keeps revealing itself I am tempted to just replace it altogether by our own code that parses only what we want it to parse, and doesn't eat whitspace and so on. it feels right now we need more code to block all the stuff it does we don't want than it would take to write our own implementation that only does what we want. but maybe let's leave that for a later PR)

**Comment by poettering** [comment]
force pushed a new version. added the test case, but not much else, see comment above

File: src/basic/parse-util.c (hunk 7 above)
Comment (keszybz) [reply src/basic/parse-util.c:439]: > replace it altogether by our own code that parses only what we want it to parse

Indeed.

> let's leave that for a later PR

Ack.

**Comment by keszybz** [comment]
LGTM.

File: src/basic/parse-util.c [hunk 9]
```diff
@@ -389,13 +430,17 @@ int safe_atou_full(const char *s, unsigned base, unsigned *ret_u) {
 }
```
Comment (ripplewang888) [inline src/basic/parse-util.c:430]: 1
